### PR TITLE
Supervisor metadata auto cleanup failing as missing Guice injection

### PR DESCRIPTION
**Pull request by maytasm**
Supervisor metadata auto cleanup failing as missing Guice injection

### Description

The Supervisor auto cleanup feature fails to initialize KafkaSupervisorSpec as the TaskStorage injection is missing. The TaskStorage is bind in Overlord but not Coordinator.

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: services/src/main/java/org/apache/druid/cli/CliCoordinator.java [hunk 1]
```diff
@@ -289,6 +291,9 @@ public void configure(Binder binder)
 
             Jerseys.addResource(binder, SelfDiscoveryResource.class);
             LifecycleModule.registerKey(binder, Key.get(SelfDiscoveryResource.class));
+
+            // This is needed to deserialized SupervisorSpec for Supervisor Auto Cleanup
```
Comment (clintropolis) [inline services/src/main/java/org/apache/druid/cli/CliCoordinator.java:295]: nit:
```suggestion
            // This is needed to deserialize SupervisorSpec for Supervisor Auto Cleanup
```

**Review by clintropolis — APPROVED** [review]
:+1: after ci

File: services/src/main/java/org/apache/druid/cli/CliCoordinator.java (hunk 1 above)
Comment (maytasm) [reply services/src/main/java/org/apache/druid/cli/CliCoordinator.java]: Done

**Review by abhishekagarwal87 — APPROVED** [review]

**Comment by maytasm** [comment]
@clintropolis @abhishekagarwal87 Sorry to bother you both but I had to make some changes to the PR. Turns out that I cannot bind TaskStorage in CliCoordinator. This is because when we run the Coordinator together with the Overlord, then both CliCoordinator and CliOverlord will both try to bind TaskStorage, resulting in Guice error. I found out that you can annotate using the @Nullable to tell Guice to bind null if the binding is not given. 

Hence, removed `binder.bind(TaskStorage.class).toProvider(Providers.of(null));` in CliCoordinator and added @Nullable  to the constructor of the different SupervisorSpecs.

**Review by clintropolis — APPROVED** [review]

**Comment by clintropolis** [comment]
>Sorry to bother you both but I had to make some changes to the PR. Turns out that I cannot simply bind TaskStorage in CliCoordinator. This is because when we run the Coordinator together with the Overlord, then both CliCoordinator and CliOverlord will both try to bind TaskStorage, resulting in Guice error.

Oops, good catch :+1:

**Pull request merged**